File: pnix-rs/pnix-rs/src/production_outcome.rs
use std::fs::File;
use std::io::{BufRead, BufReader};

use crate::machine_outcome::{EvalError, MachineOutcome, SCHEMA as HOST_OUTCOME_SCHEMA};
use crate::px;

const PROJECTION_SCHEMA: &str = "pnix.production-basic-outcome-projection.v1";

fn failed(phase: &str, class: &str) -> MachineOutcome<px::PxVal> {
    MachineOutcome::Failed(EvalError {
        phase: String::from(phase),
        class: String::from(class),
        evidence: std::collections::BTreeMap::new(),
    })
}

pub fn eval_source_outcome(source: &str) -> MachineOutcome<px::PxVal> {
    match px::px_run_value_outcome(source)
        .and_then(|value| px::px_force_deep_outcome(&value))
    {
        Ok(value) => MachineOutcome::Done(value),
        Err(error) => failed(error.phase_name(), error.class_name()),
    }
}

#[derive(Clone)]
struct Projection {
    outcome_kind: &'static str,
    error_phase: Option<String>,
    error_class: Option<String>,
    value_json: Option<String>,
}

fn project(outcome: &MachineOutcome<px::PxVal>) -> Projection {
    match outcome {
        MachineOutcome::Done(value) => Projection {
            outcome_kind: "done",
            error_phase: None,
            error_class: None,
            value_json: Some(px::px_to_json(value).unwrap_or_else(|_| String::from("null"))),
        },
        MachineOutcome::Failed(error) => Projection {
            outcome_kind: "failed",
            error_phase: Some(error.phase.clone()),
            error_class: Some(error.class.clone()),
            value_json: None,
        },
        _ => panic!("basic projection accepts Done or Failed"),
    }
}

fn json_quote(value: &str) -> String {
    let mut output = String::from("\"");
    for character in value.chars() {
        match character {
            '"' => output.push_str("\\\""),
            '\\' => output.push_str("\\\\"),
            '\n' => output.push_str("\\n"),
            '\r' => output.push_str("\\r"),
            '\t' => output.push_str("\\t"),
            other => output.push(other),
        }
    }
    output.push('"');
    output
}

fn option_json(value: &Option<String>) -> String {
    match value {
        Some(item) => json_quote(item),
        None => String::from("null"),
    }
}

impl Projection {
    fn to_json(&self) -> String {
        format!(
            "{{\"error_class\":{},\"error_phase\":{},\"outcome_kind\":{},\"schema\":{},\"value_json\":{}}}",
            option_json(&self.error_class),
            option_json(&self.error_phase),
            json_quote(self.outcome_kind),
            json_quote(PROJECTION_SCHEMA),
            option_json(&self.value_json),
        )
    }
}

fn expected_projection(kind: &str, phase: &str, class: &str, value: &str) -> Projection {
    Projection {
        outcome_kind: if kind == "done" { "done" } else { "failed" },
        error_phase: if phase.is_empty() { None } else { Some(String::from(phase)) },
        error_class: if class.is_empty() { None } else { Some(String::from(class)) },
        value_json: if value.is_empty() { None } else { Some(String::from(value)) },
    }
}

pub fn report_json(path: &str) -> Result<(String, bool), String> {
    let file = File::open(path).map_err(|error| error.to_string())?;
    let mut rows = Vec::new();
    let mut all_match = true;
    for line in BufReader::new(file).lines() {
        let line = line.map_err(|error| error.to_string())?;
        if line.is_empty() {
            continue;
        }
        let fields: Vec<&str> = line.splitn(6, '\t').collect();
        if fields.len() != 6 {
            return Err(String::from("invalid basic outcome case row"));
        }
        let projection = project(&eval_source_outcome(fields[5]));
        let expected = expected_projection(fields[1], fields[2], fields[3], fields[4]);
        let matches = projection.to_json() == expected.to_json();
        all_match = all_match && matches;
        rows.push(format!(
            "{{\"case\":{},\"matches_expected\":{},\"projection\":{}}}",
            json_quote(fields[0]),
            if matches { "true" } else { "false" },
            projection.to_json(),
        ));
    }
    Ok((format!(
        "{{\"host\":\"pnix-rs\",\"host_outcome_schema\":{},\"matrix\":[{}],\"model_schema\":\"pnix.machine.eval-outcome-model.v1\",\"schema\":\"pnix.production-basic-outcome-report.v1\",\"status\":{{\"automatic_codegen\":false,\"basic_language_errors_are_held\":false,\"legacy_error_transport_is_semantic_owner\":false,\"production_basic_outcome_convergence_v1\":true,\"production_common_machine_replacement\":false,\"production_requested_integration\":false,\"production_suspension_equivalence\":false}}}}",
        json_quote(HOST_OUTCOME_SCHEMA),
        rows.join(","),
    ), all_match))
}

File: pnix-clr/pnix-clr/src/pnix_clr/production_outcome.clj
(ns pnix-clr.production-outcome
  "ClojureCLR projection of the common production basic-outcome contract.
  `--self-check` verifies this host's own nominal outcome boundary in
  isolation. `production-report` can additionally replay an externally
  supplied case-matrix TSV (case/kind/phase/class/expected-value/source
  columns) against this host's evaluator; it is not wired to any sibling
  corpus tree by default and supplies only the CLR runtime/observer adapter."
  (:require [clojure.string :as str]
            [pnix-clr.evaluator :as evaluator]
            [pnix-clr.host :as host]
            [pnix-clr.json :as json]
            [pnix-clr.outcome :as outcome]))

(def report-schema "pnix.production-basic-outcome-report.v1")

(defn project-production-outcome
  [result]
  (cond
    (outcome/done? result)
    (sorted-map
     "error_class" nil
     "error_phase" nil
     "outcome_kind" "done"
     "schema" outcome/production-projection-schema
     "value_json" (json/write-json (outcome/value-of result)))

    (outcome/failed? result)
    (let [error (outcome/error-of result)]
      (sorted-map
       "error_class" (:class error)
       "error_phase" (:phase error)
       "outcome_kind" "failed"
       "schema" outcome/production-projection-schema
       "value_json" nil))

    :else
    (throw (ex-info "basic projection accepts Done or Failed" {}))))

(defn- optional-field
  [value]
  (when-not (str/blank? value) value))

(defn- expected-projection
  [kind phase class-name value-json]
  (sorted-map
   "error_class" (optional-field class-name)
   "error_phase" (optional-field phase)
   "outcome_kind" kind
   "schema" outcome/production-projection-schema
   "value_json" (optional-field value-json)))

(defn- read-cases
  [path]
  (mapv #(str/split % #"\t" 6)
        (remove str/blank?
                (str/split-lines
                 (System.IO.File/ReadAllText
                  (System.IO.Path/GetFullPath (str path)))))))

(defn production-report
  [path]
  (let [matrix
        (mapv
         (fn [[case-name kind phase class-name expected-value source]]
           (let [projection
                 (project-production-outcome
                  (evaluator/eval-source source))
                 expected
                 (expected-projection
                  kind phase class-name expected-value)]
             (sorted-map
              "case" case-name
              "matches_expected" (= expected projection)
              "projection" projection)))
         (read-cases path))
        all-ok (every? #(get % "matches_expected") matrix)]
    (sorted-map
     "all_ok" all-ok
     "host" "pnix-clr"
     "host_outcome_schema" outcome/schema
     "matrix" matrix
     "model_schema" outcome/model-schema
     "schema" report-schema
     "status"
     (sorted-map
      "automatic_codegen" false
      "basic_language_errors_are_held" false
      "legacy_error_transport_is_semantic_owner" false
      "production_basic_outcome_convergence_v1" all-ok
      "production_common_machine_replacement" false
      "production_requested_integration" false
      "production_suspension_equivalence" false))))

(defn- usage!
  []
  (binding [*out* *err*]
    (println
     "usage: pnix-clr.production-outcome --self-check | CASES.tsv"))
  (host/exit! 2))

(defn -main
  [& args]
  (cond
    (= ["--self-check"] (vec args))
    (println (json/write-json (outcome/self-check)))

    (= 1 (count args))
    (let [report (production-report (first args))]
      (println (json/write-json report))
      (when-not (get report "all_ok")
        (host/exit! 1)))

    :else
    (usage!)))
